### DISCUSSION TITLE
Invalid CSRF token error with VueJs frontend

### DISCUSSION THREAD

**Discussion by Jumborner** [symfony/symfony#42437] [Q&A]
Hello. I am having trouble in making authentication work using an external frontend ( vue ) with my symfony app. The main problem is the "Invalid CSRF token" error. I have a login form in vue which sends an object containing the username, password, and the csrf token ( which I get from symfony tokengenerator ). I have a custom authenticator where I create the user passport and add the token to it. 
`public function authenticate(Request $request): PassportInterface
   {
       $username = $request->request->get('username', '');
       $request->getSession()->set(Security::LAST_USERNAME, $username);
       $passport = new Passport(
           new UserBadge($username),
           new PasswordCredentials($request->request->get('password', '')),
       );
       $passport->addBadge(new CsrfTokenBadge('authenticate', $request->get('_csrf_token')));
       return $passport;
   }`

It makes it through to the AuthenticationManager.php, where it enters the executeAuthenticator method. The error comes after the CheckPassportEvent is dispatched, from CSRFProtectionListener. It fails on the `if (false === $this->csrfTokenManager->isTokenValid($csrfToken))`.

I have tried to get the tokenmanager instance inside of my authenticator and create the token there and add it to the passport.
`$token = $this->csrfTokenManager->getToken('authenticate');
        $passport->addBadge(new CsrfTokenBadge($token->getId(), $token->getValue()));`
This lets me past the authentication, but immediately afterwards, when it redirects to the next path, it gives me an error "Access denied, the user is not fully authenticated; redirecting to authentication entry point.". After some debugging, it seems that the token storage is empty ( the token is saved to the storage when the getToken() method is called ).

When I do the authentication with the twig template, it works flawlessly. How exactly {{ csrf_token('authenticate') }} makes and handles the token I do not understand. Any input would be appreciated.

P.S GitHub's "insert code" doesn't seem to format it correctly =/ 

Bump. Been stuck on this issue for 2 days now. No clue how to make proper authentication with the new authentication system and separate VueJs frontend.

After tracing some of the calls, it seems it tries to get the session from the request? How is it supposed to work? =/

**Comment by dbrumann** [comment]
FYI: You can format the code block using \`\`\` and then closing it again with \`\`\`.

When you call `isTokenValid($token)` it will compare the provided token with the one generated when calling `$tokenManager->getToken('authorize')` that is stored in the session, as you already noted.

My guess is, that either a different request was made, generating a new token and storing it in the session, so it no longer matches the one you have or that the token is not in the storage. This can happen, if you run the application behind a load balancer and requests go to different instances, that do not have a shared session storage (alternatively session pinning can be used).

The only advice I can give is to check if the session id stays the same between generating the login form and submitting it. Checking if the stored token changes between requests can be tricky. Maybe you can use a file watcher on the session file (if you use native file storage) to check if it doesn't change between requests.

**Reply by Jumborner** [reply]
After succeeding in the login, whenever I send a request to a protected route, I get an error " User not fully authorized". After debugging, it seems that it cant get the looged in user from the session because the session does not exist. Inside of the "ContextListener.php", the session is set to null at ``` $session = $request->hasPreviousSession() && $request->hasSession() ? $request->getSession() : null; ```

It tries to get the session from the request? I'm supposed to pass it with each request? =/

I am new to symfony and especially to the new authentication method. I could not find any relevant examples solving this issue =/

**Comment by idybil** [comment]
Hi, already had a problem like this in the past, the application was sending multiple ajax requests which regenerate the session id. Can you test by adding the next configuration in the security.yaml file :

security:
    # Do not regenerate session id on each request.
    session_fixation_strategy: none
    ...


Not sure this is the best way to fix your problem (if it works for you).  On my side I didn't care about security since the application was a intranet.

**Reply by Jumborner** [reply]
Alas, this seems not to be the case
